src/Data/Monoid/Average.hs

```diff
@@ -25,101 +24,40 @@ module Data.Monoid.Average (
 
-import Data.Typeable
+import Control.Applicative
+import Control.Monad
+import Data.AdditiveGroup
 import Data.Maybe
 import Data.Semigroup
-import Data.AdditiveGroup
+import Data.Typeable
 import Data.VectorSpace
-import Data.AffineSpace
-import Control.Monad
-import Control.Applicative
 
 -- |
 -- A monoid for 'Average' values.
 --
--- This is actually just the free monoid with an extra function 'average' for
--- extracing the (arithmetic) mean. This function is used to implement 'Real',
--- so you can use 'Average' whenever a ('Monoid', 'Real') is required.
---
--- >>> toRational $ mconcat [1,2::Average Rational]
--- 3 % 2
--- >>> toRational $ mconcat [1,2::Sum Rational]
--- 3 % 1
--- >>> toRational $ mconcat [1,2::Product Rational]
--- 2 % 1
+-- This average encapsulates length and sum in a space efficient form.
 --
-newtype Average a = Average { getAverage :: [a] }
-  deriving (Show, Semigroup, Monoid, Typeable, Functor, Applicative)
+data Average a = Average !Int a
+  deriving (Show, Eq, Ord, Typeable, Functor)
 
-instance (Fractional a, Eq a) => Eq (Average a) where
-  a == b = average a == average b
+averageDatum :: a -> Average a
+averageDatum = Average 1
 
-instance (Fractional a, Ord a) => Ord (Average a) where
-  a `compare` b = average a `compare` average b
-  
--- What should (+) and (*) do for Average values?
--- 
--- The important thing is to preserve scalar addition and multiplication (for example
--- scaling all components of) an average value by some constant factor, so we can just as
--- well use the standard list instance. What about averages with more components? I *think*
--- 'average' is a linear map, so they would work as expected:
--- 
--- >>> average (2<>2<>3)+average (3<>3)
--- 16 % 3
--- >>> average $ (2<>2<>3)+(3<>3)
--- 16 % 3
--- >>> average (mconcat [5,6,9])*average (mconcat[-1,0])
--- (-10) % 3
--- >>> average $ (mconcat [5,6,9])*(mconcat[-1,0])
--- (-10) % 3
--- 
+instance Num n => Semigroup (Average n) where
+  Average lx nx <> Average ly ny = Average (lx + ly) (nx + ny)
 
-instance Num a => Num (Average a) where
-  (+) = liftA2 (+)
-  (*) = liftA2 (*)
-  negate = fmap negate
-  abs    = fmap abs
-  signum = fmap signum
-  fromInteger = pure . fromInteger
-  
-instance (Fractional a, Num a) => Fractional (Average a) where
-  (/) = liftA2 (/)
-  fromRational = pure . fromRational
+instance Num n => Monoid (Average n) where
+  mappend = (<>)
+  mempty = Average 0 0
 
-instance (Real a, Fractional a) => Real (Average a) where
-  toRational = toRational . average
-
-instance Floating a => Floating (Average a) where
-  pi = pure pi
-  exp = fmap exp
-  sqrt = fmap sqrt
-  log = fmap log
-  sin = fmap sin
-  tan = fmap tan
-  cos = fmap cos
-  asin = fmap asin
-  atan = fmap atan
-  acos = fmap acos
-  sinh = fmap sinh
-  tanh = fmap tanh
-  cosh = fmap cosh
-  asinh = fmap asinh
-  atanh = fmap atanh
-  acosh = fmap acosh
-  
 instance AdditiveGroup a => AdditiveGroup (Average a) where
```

Comment:
This `AdditiveGroup` instance has a very different behaviour from the old one, and IMO doesn't make sense. (In the old instance, `zeroV` would represent zero with weight one. Likewise, numerical literals represented their value with weight one, so you could do `3 <> 5` and something going for _the average value_, namely `Average [3,5]` which represents `4` via `maybeAverage`.)

What you're proposing not only sacrifices this handy behaviour, it also violates the vector space laws, because e.g. ```
Average 1 1 ^+^ Average 1 1 ≡ Average 1 1
```
but
```
2 *^ Average 1 1 ≡ Average 1 2
```

Reply:
You are right about violating the vector space laws. I've added quick check tests for these and `distributive: c d v` is currently failing. However `zeroV = Average 1 zeroV` is also problematic as it violates identity laws.

I'm also not sold that a `Num` instance makes any sense. It is very convenient to use numeric literals, but I'm not sure the existing implementation of `*` was coherent or useful.

Reply:
Yeah, the `Num` instance was definitely always a bit funny. I'm not at all sure it's _possible_ to make it obey all the laws, however the original list-based implementation at least behaved properly when doing arithmetic on number literals (i.e. on weight-1 numbers). To restore that behaviour with the new implementation, we **1.** _multiply_ weights when `ap`ing values (as I proposed earlier), not add them, and **2.** compute mappings on the _averaged result_, i.e. basically
```
fmap f (Average w x) = Average w $ w' * f (x / w')
 where w' = fromIntegral w
```
And there our `Functor` and `Applicative` instances are dead, because this only works when the contained value is actually vector space. We could add best make it a [`MonoFunctor`](http://hackage.haskell.org/package/mono-traversable-1.0.2.1/docs/Data-MonoTraversable.html#t:MonoFunctor) or [constrained `Applicative`](http://hackage.haskell.org/package/constrained-categories-0.3.0.1/docs/Control-Applicative-Constrained.html).

But as you said, perhaps we'd better just not define any functor instances at all. We could still do
```
Average wx x ^+^ Average wy y = Average wxy $ fromIntegral wxy
                    *^ (x ^/ fromIntegral wx ^+^ y ^/ fromIntegral wy)
 where wxy = wx * wy
```